### DISCUSSION TITLE
HA su docker windows - Problema Caddy e Reverse Proxy

### DISCUSSION THREAD

**Discussion by omarantelmi** [indomus/forum#1641] [HUB: Home Assistant]
Salve, spero di fornire tutte le info utili per potervi permettere di aiutarmi in merito a questo problema

La mia configurazione attuale include un pc portatile Windows con Docker, vorrei cercare di abilitare HTTPS tramite Reverse Proxy e Caddy

Sto seguendo questa guida: https://indomus.it/guide/installare-e-configurare-caddy-reverse-proxy-con-docker-su-linux-debian/

Ho istanziato correttamente Caddy su Docker, il problema sembrano essere i certificati che dovrebbero essere scaricati e inseriti in C:\Users\MacRapSERVER\caddy\data ma questo non avviene, la cartella risulta vuota

Invio il log di Caddy specificando però che ho utilizzato la porta 443 (Non so perchè nel log mi indica la porta 80)

```
{"level":"info","ts":1650101956.9967666,"msg":"using provided configuration","config_file":"/etc/caddy/Caddyfile","config_adapter":"caddyfile"}

{"level":"info","ts":1650101957.0078187,"logger":"admin","msg":"admin endpoint started","address":"tcp/localhost:2019","enforce_origin":false,"origins":["localhost:2019","[::1]:2019","127.0.0.1:2019"]}

{"level":"info","ts":1650101957.0099368,"logger":"http","msg":"server is listening only on the HTTP port, so no automatic HTTPS will be applied to this server","server_name":"srv0","http_port":80}

{"level":"info","ts":1650101957.0106127,"logger":"tls.cache.maintenance","msg":"started background certificate maintenance","cache":"0xc0002c8000"}

{"level":"info","ts":1650101957.0110679,"logger":"tls","msg":"cleaning storage unit","description":"FileStorage:/data/caddy"}

{"level":"info","ts":1650101957.0128763,"msg":"autosaved config (load with --resume flag)","file":"/config/caddy/autosave.json"}

{"level":"info","ts":1650101957.0129118,"msg":"serving initial configuration"}

{"level":"info","ts":1650101957.0135624,"logger":"tls","msg":"finished cleaning storage units"}
```

Vi ringrazio in anticipo

**Comment by snakuzzo** [comment]
Se leggi il log c'è scritto...

```
server is listening only on the HTTP port, so no automatic HTTPS will be applied to this server
```

Solo con questo log non ci facciamo niente.
Devi condividere tutto...

- come hai creato il container
- il Caddyfile
- dove hai creato i file

**Reply by omarantelmi** [reply]
Ok elenco i passaggi

1) Apertura Porta Router 443 verso IP HA
2) Creazione della cartella C:\Users\MacRapSERVER\caddy\data (Non so se ho fatto bene, ho creato le cartelle manualmente, prima cartella caddy e poi data)
3) Creo il file Caddyfile (senza estensione) con le seguenti righe e lo posiziono in C:\Users\MacRapSERVER\caddy
```
#  When your config is ok comment block below to disable use of AMCE staging server
{
    acme_ca https://acme-staging-v02.api.letsencrypt.org/directory 
}

https://PERSONALIZZATO.duckdns.org { 
  respond "Saluti dal Crew inDomus.it"
  file_server
}
```
4) Eseguo il comando per istanziare Caddy, quindi CDM e digito
`docker run --init -d --name="caddy" -e "TZ=Europe/Rome" -v //c/Users/MacRapSERVER/caddy/Caddyfile:Caddyfile -v //c/Users/MacRapSERVER/caddy/data:/data --restart always -p 443:443 caddy:latest`


Ora qui vado a verificare che si siano creati effettivamente i certificati, e i certificati non ci sono. Credo che il comando su sia errato, ma non riesco a capire come sistemarlo

Il resto non dovrebbe essere un problema perchè si tratta essenzialmente di modificare il caddyfile e ricreare il container

**Reply by snakuzzo** [reply]
Di sicuro sbagli il mount del `Caddyfile`.
Il mount prevede `<percorso_host>:<precorso_container>`. Quindi...visto che di default si aspetta che il file sia `/etc/caddy/Caddyfile`, il comando per la creazione del container andrebbe cambiato...

```
docker run --init -d --name="caddy" -e "TZ=Europe/Rome" -v //c/Users/MacRapSERVER/caddy/Caddyfile:/etc/caddy/Caddyfile -v //c/Users/MacRapSERVER/caddy/data:/data --restart always -p 443:443 caddy:latest
```

Ad ogni modo puoi controllare queste cose sulla documentazione ufficiale di Caddy (su github o su dockerhub). Una volta creato il container puoi anche "entrare" (concedimi il termine) nel container per controllare che il file sia stato mappato correttamente...

```
docker exec -it <nome_container> sh
/srv # cat /etc/caddy/Caddyfile
```

**Reply by omarantelmi** [reply]
Il problema era il comando errato, il quale non faceva creare i certificati, confermo la risoluzione del problema
Grazie mille
